scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala

```diff
@@ -113,18 +97,57 @@ object Grouped {
    * If we are using OrderedComparable, we need to box the key
    * to prevent other serializers from handling the key
    */
-  def tuple2Setter[K, V](ord: Ordering[K]): TupleSetter[(K, V)] =
-    ord match {
-      case _: OrderedSerialization[_] =>
-        tup2Setter[(BoxedKey[K], V)].contraMap { kv1: (K, V) =>
-          (BoxedKey(kv1._1), kv1._2)
+  def maybeBox[K, V](ord: Ordering[K])(op: (TupleSetter[(K, V)], Fields) => Pipe): Pipe = ord match {
+    case ordser: OrderedSerialization[K] =>
+      val (boxfn, cls) = Boxed.next[K]
+      val ts = tup2Setter[(Boxed[K], V)].contraMap { kv1: (K, V) => (boxfn(kv1._1), kv1._2) }
+      val boxordSer = BoxedOrderedSerialization(boxfn, ordser)
+      val keyF = new Fields("key")
+      keyF.setComparator("key", new CascadingBinaryComparator(boxordSer))
+      val pipe = op(ts, keyF)
+
+      case class ToVisit[T](queue: Queue[T], inQueue: Set[T]) {
+        def maybeAdd(t: T): ToVisit[T] = if (inQueue(t)) this else {
+          ToVisit(queue :+ t, inQueue + t)
         }
-      case _ => tup2Setter[(K, V)]
-    }
+        def next: Option[(T, ToVisit[T])] =
+          if (inQueue.isEmpty) None
+          else Some((queue.head, ToVisit(queue.tail, inQueue - queue.head)))
+      }
+
+      @annotation.tailrec
+      def go(p: Pipe, visited: Set[Pipe], toVisit: ToVisit[Pipe]): Set[Pipe] = {
```

Comment:
comment to the loop here: it is a search of all parents of this.

One note, we actually have a much simpler version of this in RichPipe, I remembered, that we use in the REPL. (.upstreamPipes or something). It uses the fact that we have a DAG so we don't need the most general algorithm you have here.

We should either push this algo there, or use the existing one.